### PR TITLE
doc: SPO set up instruction

### PR DESCRIPTION
**Pull request by xiao-lix**
1. Add README.md file with instructions to create SPFx extension solution installing AppInslights JS SDK.
2. Add AppInsightsExtensionSolutionSample - a SPFx extension solution sample.

### CODE REVIEW

File: SPO/AppInsightsExtensionSolutionSample/.editorconfig [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# EditorConfig helps developers define and maintain consistent
+# coding styles between different editors and IDEs
+# editorconfig.org
+
+root = true
+
+
```
Comment (markwolff) [inline SPO/AppInsightsExtensionSolutionSample/.editorconfig:7]: nit: delete dupe newline

File: SPO/AppInsightsExtensionSolutionSample/.vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      "request": "launch",
+      "url": "https://enter-your-SharePoint-site/_layouts/workbench.aspx",
+      "webRoot": "${workspaceRoot}",
+      "sourceMaps": true,
+      "sourceMapPathOverrides": {
+        "webpack:///.././src/*": "${webRoot}/src/*",
+        "webpack:///../../../src/*": "${webRoot}/src/*",
+        "webpack:///../../../../src/*": "${webRoot}/src/*",
+        "webpack:///../../../../../src/*": "${webRoot}/src/*"
+      },
```
Comment (markwolff) [inline SPO/AppInsightsExtensionSolutionSample/.vscode/launch.json:36]: nice!

File: SPO/AppInsightsExtensionSolutionSample/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
+## appinsights-extension-solution-sample
+
+This is where you include your WebPart documentation.
```
Comment (markwolff) [inline SPO/AppInsightsExtensionSolutionSample/README.md:3]: add description

File: SPO/AppInsightsExtensionSolutionSample/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+### Building the code
+
+```bash
+git clone the repo
+npm i
+npm i -g gulp
+gulp
+```
```
Comment (markwolff) [inline SPO/AppInsightsExtensionSolutionSample/README.md:12]: This needs to be updated for being in this monorepo

File: SPO/README.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
+# How to create a SPFx extension solution with AppInsights installed from scratch
```
Comment (xiao-lix) [inline SPO/README.md:1]: @DawgFan hi Matt, can you please help check the language and pictures in README? Thanks! 😁

File: SPO/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+        <ClientSideComponentInstance
+            Title="AppinsightsExtension"
+            Location="ClientSideExtension.ApplicationCustomizer"
+            ComponentId="96262ed0-fe5d-4786-abc2-9f6b9cb3fefc"
+            Properties="{&quot;instrumentationKey&quot;:&quot;YOUR_INSTRUMENTATION_KEY_GOES_HERE&quot;}">
+        </ClientSideComponentInstance>
+    </Elements>
+    ```
+    
+3. Run following command to create .sppkg file. Check [Host extension from CDN](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/extensions/get-started/hosting-extension-from-office365-cdn) for reference.
```
Comment (DawgFan) [inline SPO/README.md:55]: This is a lot of steps. What do you think about creating a SKPPG file for OOB enablement, then reference full doc for anyone who wants custom? I don't want to create more to maintain, but it also seems like some customers may give up.

File: SPO/README.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+    const appInsights = new ApplicationInsights({ config: {
+    instrumentationKey: this.properties.instrumentationKey
+    /* ...Other Configuration Options... */
+    } });
+    appInsights.loadAppInsights();
+    appInsights.trackPageView(); // Manually call trackPageView to establish the current user/session/pageview
+    ```
+## Set up deployment configuration
+### Deploy Extension solution Tenant wide (All sites share same iKey)
+1. Under your Extension solution config folder, set the skipFeatureDeployment attribute to true in the package-solution.json file. And add "ClientSideInstance.xml" file under "features"/"assets"/"elementManifests".
```
Comment (DawgFan) [inline SPO/README.md:39]: Consider continuing the numbers if this is the next step. You may need to have two forks - Option A: 4,5,6... and OPTIONB 4,5,6...

File: SPO/README.md (hunk 5 above)
Comment (DawgFan) [reply SPO/README.md:1]: Thank you for adding all the great screen shots!

File: SPO/AppInsightsExtensionSolutionSample/README.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
+## appinsights-extension-solution-sample
+
+This is a sample SharePoint Framework Extension solution that will set up Application Insights on a SharePoint Online modern site.
```
[the remark this answers is not in the document]
Comment (markwolff) [reply SPO/AppInsightsExtensionSolutionSample/README.md:3]: Yes this looks good. You could add a link to [sharepoint online modern](https://support.office.com/en-IE/Article/sharepoint-classic-and-modern-experiences-5725c103-505d-4a6e-9350-300d3ec7d73f) here.

File: SPO/AppInsightsExtensionSolutionSample/README.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+```bash
+git clone the repo
+```
+Inside AppInsightsExtensionSolutionSample root folder
+```bash
+npm i
+npm i -g gulp
+gulp
+```
+For solution deployment, check out the [Set up deployment](https://github.com/microsoft/ApplicationInsights-JS/tree/master/SPO/README.md#set-up-deployment) configuration section.
```
[the remark this answers is not in the document]
Comment (markwolff) [reply SPO/AppInsightsExtensionSolutionSample/README.md:16]: This works. Am I correct that in order to deploy this, a user needs to go and build it first? Is there some prebuilt thing they can grab somewhere, somehow? Or are these just instructions to build this package to contribute to it?

[the remark this answers is not in the document]
Comment (xiao-lix) [reply SPO/AppInsightsExtensionSolutionSample/README.md:16]: This is just the instruction to build this solution. `npm i` will install its dependencies and AppInsights SDK, `npm i -g gulp` is to make sure user installs `gulp` in order use gulp command. To deploy they'll need to follow this : https://docs.microsoft.com/en-us/sharepoint/dev/spfx/extensions/get-started/serving-your-extension-from-sharepoint#deploy-the-extension-to-sharepoint-online-and-host-javascript-from-local-host

File: SPO/README.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+4. Open your tenant app catalog.
+5. Click on Apps for SharePoint.
+6. Upload the solution to the catalog.
+7. Click Deploy.
+<p><img src="./img/image1.png"/></p>
+8. On the site you want to install Application Insights, click "+New" -> App , to add appinsights extension to your site.
+> Wait until the extension is added to your site, AppInsights Web JS SDK is installed on your site. With this route, you can not edit properties on SharePoint UI. If you change the AI configuration, you'll need to repeate step 3-8.
+<p><img src="./img/image2.png"/></p>
+
+Check [AppInsightsExtensionSolutionSample](https://github.com/microsoft/ApplicationInsights-JS/tree/master/SPO/AppInsightsExtensionSolutionSample).
```
[the remark this answers is not in the document]
Comment (xiao-lix) [reply SPO/README.md:103]: Yes this link should work after the PR gets merged. :)

File: SPO/README.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    gulp bundle --ship
+    gulp package-solution --ship
+    ```
+4. Open your tenant app catalog.
+5. Click on Apps for SharePoint.
+6. Upload the solution to the catalog.
+7. Check the option Make this solution available to all sites in the organization .
+<p><img src="./img/image4.png"/></p>
+
+> Now Application Insights web JS SDK is installed tenant wide on your SPO. If you need to edit configurations, you could do it through SharePoint UI:
```
[the remark this answers is not in the document]
Comment (xiao-lix) [reply SPO/README.md:66]: yes, sure, I'll add it.

File: SPO/README.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
+# How to create a SPFx extension solution with AppInsights installed from scratch
+
+## Set up Application Insights
```
[the remark this answers is not in the document]
Comment (xiao-lix) [reply SPO/README.md:3]: I'm not sure how much we should put on official docs. The problem is we have no way to edit UI to add AppInsights as the old doc does. If a customer wants to create his own solution, he'll have to follow all steps listed here. I didn't list local build though... that will be longer. If this is too long, we can cut the deployment to single site part on official doc? At the end point them to github? Or we have to upload a sppkg file somewhere for users to download? We decided not to do that right?

File: SPO/README.md (hunk 6 above)
Comment (xiao-lix) [reply SPO/README.md:55]: Yeah they have many moving parts. SPO have 3 files for the same properties... (where you link the iKey and other configs to your solution):

- If you build locally, edit `serve.json` -> `properties` (I didn't add this part in README)
- If you deploy to single site, edit `element.xml` -> `ClientSideComponentProperties` 
- If you deploy tenant wide, edit `ClientSideInstance.xml` -> `Properties` 

And they are under different folders. The only way I can think of is to upload sppkg file for them to download, but in that way user will lose the flexibility to config appInsights and add their custom data, which will still raise supportability I guess.

File: SPO/README.md (hunk 7 above)
Comment (xiao-lix) [reply SPO/README.md]: Sounds good, I'll update it thanks.

File: SPO/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (DawgFan) [reply SPO/README.md:6]: As a starting point, we can link out to GitHub from official docs.

I wish we could provide a sppkg download from official docs for the short/easy version -- yes we decided to not do it...but it just feels like the Cx experience could be better.

Let's get this version out on Github, then I can revisit with Ram.

File: SPO/README.md (hunk 6 above)
Comment (DawgFan) [reply SPO/README.md]: Thanks Xiao. I actually like your meta explanation above. Should we work this in to docs?

Comment (xiao-lix) [reply SPO/README.md]: Yeah that sounds reasonable. I'll add it and will tag you :) Will need your help to address language.

File: SPO/README.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    appInsights.trackPageView(); // Manually call trackPageView to establish the current user/session/pageview
+    ```
+## Set up deployment configuration
+### Select Option A or Option B for deployment based on your needs: 
+The custom properties (eg. instrumentationKey in this example) is stored in several locations: 
+1. When running a gulp serve and running the page in debug mode, the key that is stored in `config\serve.json` -> `properties` is passed to the page as querystring.
+2. When the extension is installed on a site, the key in `sharepoint\assets\elements.xml` -> `ClientSideComponentProperties` is used (Option B).
+3. When deployed tenant wide, the key in `sharepoint\assets\ClientSideInstance.xml` -> `Properties` is used (Option A).
+
+The [AppInsightsExtensionSolutionSample](https://github.com/microsoft/ApplicationInsights-JS/tree/master/SPO/AppInsightsExtensionSolutionSample) is using Option A. We highly recommend you to create your own SPFx Extension solution, in order to get the full support of Application Insights and configure the SDK based on your needs.
```
Comment (xiao-lix) [inline SPO/README.md:48]: @DawgFan hi Matt, do you think this section works?

File: SPO/README.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    /* ...Other Configuration Options... */
+    } });
+    appInsights.loadAppInsights();
+    appInsights.trackPageView(); // Manually call trackPageView to establish the current user/session/pageview
+    ```
+## Set up deployment configuration
+### Select Option A or Option B for deployment based on your needs: 
+The custom properties (eg. instrumentationKey in this example) is stored in several locations: 
+1. When running a gulp serve and running the page in debug mode, the key that is stored in `config\serve.json` -> `properties` is passed to the page as querystring.
+2. When the extension is installed on a site, the key in `sharepoint\assets\elements.xml` -> `ClientSideComponentProperties` is used (Option B).
```
Comment (DawgFan) [inline SPO/README.md:45]: Switch #2 & #3 so Option A, B is the same order as below.

File: SPO/README.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+> Now Application Insights web JS SDK is installed tenant wide on your SPO. If you need to edit configurations, you could do it through SharePoint UI: 
+
+11. On your tenant app catalog page, go to Site Contents.
+12. Open the list Tenant Wide Extensions.
+13. Edit the ApplicationInsightsExtension item. For example, you can use a different instrumentationKey or set other configuration properties you set up in Set up Application Insights section.
+<p><img src="./img/image5.png"/></p>
+14. Save the item and all your modern SharePoint site on Office 365 are ready to be monitored. If this was the first solution globally deployed on your tenant, it may take up to 20 minutes get available.
+<p><img src="./img/image6.png"/></p>
+
+### Option B - Deploy Extension solution on a single site
```
Comment (DawgFan) [inline SPO/README.md:87]: `### Option B - Deploy Extension solution on a single site (Individual sites may use different configurations, including iKeys)`

File: SPO/README.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+## Set up deployment configuration
+### Select Option A or Option B for deployment based on your needs: 
+The custom properties (eg. instrumentationKey in this example) is stored in several locations: 
+1. When running a gulp serve and running the page in debug mode, the key that is stored in `config\serve.json` -> `properties` is passed to the page as querystring.
+2. When the extension is installed on a site, the key in `sharepoint\assets\elements.xml` -> `ClientSideComponentProperties` is used (Option B).
+3. When deployed tenant wide, the key in `sharepoint\assets\ClientSideInstance.xml` -> `Properties` is used (Option A).
+
+The [AppInsightsExtensionSolutionSample](https://github.com/microsoft/ApplicationInsights-JS/tree/master/SPO/AppInsightsExtensionSolutionSample) is using Option A. We highly recommend you to create your own SPFx Extension solution, in order to get the full support of Application Insights and configure the SDK based on your needs.
+
+### Option A - Deploy Extension solution Tenant wide (All sites share same iKey)
```
Comment (DawgFan) [inline SPO/README.md:50]: `### Option A - Deploy Extension solution Tenant wide (All sites share same configurations, including iKey)`

File: SPO/README.md (hunk 13 above)
Comment (DawgFan) [reply SPO/README.md:48]: Yes, definitely helps! Thank you!

File: SPO/README.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+
+    const appInsights = new ApplicationInsights({ config: {
+    instrumentationKey: 'YOUR_INSTRUMENTATION_KEY_GOES_HERE'
+    /* ...Other Configuration Options... */
+    } });
+    appInsights.loadAppInsights();
+    appInsights.trackPageView(); // Manually call trackPageView to establish the current user/session/pageview
+    ```
+
+    b. Expose all configuration options you need in the ApplicationCustomizerProperties interface.
```
Comment (markwolff) [inline SPO/README.md:23]: where to find these configuration options?

**Pull request merged**